index.d.ts

```diff
@@ -0,0 +1,124 @@
+import {JsonValue} from 'type-fest';
+import {BrowserWindow, BrowserView, IpcMain, IpcRenderer} from 'electron';
+
+declare namespace ipc {
+	interface MainProcessIpc extends IpcMain {
+		/**
+		Send a message to the given window.
+
+		In the renderer process, use `ipc.answerMain` to reply to this message.
+
+		@param window - The window to send the message to.
+		@param channel - The channel to send the message on.
+		@param data - Data to send to the receiver.
+
+		@example
+		```
+		import {BrowserWindow} from 'electron';
+		import * as betterIpc from 'electron-better-ipc';
+
+		const ipc: betterIpc.MainProcessIpc = betterIpc;
+		const window = BrowserWindow.getFocusedWindow();
+
+		(async () => {
+			const emoji = await ipc.callRenderer(window!, 'get-emoji', 'unicorn');
+			console.log(emoji);
+			//=> '🦄'
+		})();
+		```
+		*/
+		callRenderer(
+			window: BrowserWindow,
+			channel: string,
+			data?: JsonValue
+		): Promise<JsonValue>;
+
+		/**
+		This method listens for a message from `ipc.callMain` defined in a renderer process and replies back.
+
+		@param channel - The channel to send the message on.
+		@param callback - The return value is sent back to the `ipc.callMain` in the renderer process.
+		@returns A function, that when called, removes the listener.
+
+		@example
+		```
+		import * as betterIpc from 'electron-better-ipc';
+
+		const ipc: betterIpc.MainProcessIpc = betterIpc;
+
+		ipc.answerRenderer('get-emoji', async emojiName => {
+			const emoji = await getEmoji(emojiName);
+			return emoji;
+		});
+		```
+		*/
+		answerRenderer(
+			channel: string,
+			callback: (
+				data: JsonValue | undefined,
```

Comment:
I'm starting to question the usefulness of the `JsonValue` type. To convert it to something unrelated, you'll have to do `data as unknown as MyType`. Any thoughts?

Reply:
Yeah, I was thinking about it the moment I wrote the definition. I would have usually used a generic type variable with a default value here but it actually doesn't make sense. Generics should only be used when a type variable is used more than once in a type signature.

I could live with `unknown` here, it's much easier to cast. I still think that `JsonValue` is the proper type here, it accurately describes the type of values to expect from `callback`.

Reply:
> I still think that JsonValue is the proper type here, it accurately describes the type of values to expect from callback.

I agree, but I also fail to see how correctness is practically useful here. What mistakes would it catch? What benefits does it bring? You would never be able to use the value directly anyway.

I notice you have changed `JsonValue` to `unknown` in some places, and left it in other places. Shouldn't we decide on either and use it consistently?

Reply:
I've changed all incoming types to `unknown` and left all outgoing ones as `JsonValue`. The reasoning for the former is ease of use. Leaving `JsonValue` for the latter allows to at least catch cases where someone would try to pass a constructor instead of an instance and similar errors.